lib/scope.js

```diff
@@ -450,13 +440,25 @@ AST_Toplevel.DEFMETHOD("mangle_names", function(options){
             to_mangle.push(node.definition());
             return;
         }
-    });
-    this.walk(tw);
+    }));
     to_mangle.forEach(function(def){ def.mangle(options) });
 
     if (options.cache) {
         options.cache.cname = this.cname;
     }
+
+    function push_symbols(dict) {
+        var a = [];
+        dict.each(function(symbol) {
+            if (options.reserved.indexOf(symbol.name) < 0) {
+                a.push(symbol);
+            }
+        });
+        a.sort(function(m, n) {
+            return n.references.length - m.references.length;
```

Comment:
I missed mentioning that within `figure_out_scope()`, `init_scope_vars()` is called on each `AST_Scope`, which resets all references to previous instances of `SymbolDef`:
https://github.com/mishoo/UglifyJS2/blob/94e5e00c0321b92aca1abf170c12a02d6c3275b5/lib/scope.js#L117
https://github.com/mishoo/UglifyJS2/blob/94e5e00c0321b92aca1abf170c12a02d6c3275b5/lib/scope.js#L262-L263